src/java/org/apache/cassandra/db/guardrails/Guardrail.java

```
@@ -0,0 +1,119 @@
+    {
+        this.name = name;
+    }
+
+    /**
+     * Checks whether this guardrail is enabled or not. This will be enabled if guardrails are
+     * ({@link Guardrails#enabled(ClientState)}) and if the authenticated user (if specified) is not system nor
+     * superuser.
+     *
+     * @param state the query state, used to skip the check if the query is internal or is done by a superuser.
+     *              A {@code null} value means that the check should be done regardless of the query.
+     * @return {@code true} if this guardrail is enabled, {@code false} otherwise.
+     */
+    public boolean enabled(@Nullable ClientState state)
+    {
+        return Guardrails.enabled(state) && (state == null || state.isOrdinaryUser());
+    }
+
+    /**
+     * Do no throw {@link InvalidRequestException} if a guardrail failure is triggered.
+     * <p>
+     * Note: this method is not thread safe and should only be used during guardrail initialization
+     *
+     * @return current guardrail
+     */
+    public Guardrail setNoExceptionOnFailure()
+    {
+        this.throwOnFailure = false;
+        return this;
+    }
+
+    protected void warn(String fullMessage, String redactedMessage)
+    {
+        logger.warn(fullMessage);
+        // Note that ClientWarn will simply ignore the message if we're not running this as part of a user query
+        // (the internal "state" will be null)
+        ClientWarn.instance.warn(fullMessage);
+        // Similarly, tracing will also ignore the message if we're not running tracing on the current thread.
+        Tracing.trace(fullMessage);
+        for (Guardrails.Listener listener : Guardrails.listeners)
+            listener.onWarningTriggered(name, redactedMessage);
+    }
+
+    protected void warn(String fullMessage)
+    {
+        warn(fullMessage, fullMessage);
+    }
+
+    protected void fail(String fullMessage, String redactedMessage)
+    {
+        logger.error(fullMessage);
```

Can we add `ClientWarn.instance.warn` as well; java/python tend to ignore exceptions msg in important cases, but will always log warnings

---

Done